### PR TITLE
MQTT plugin authenticate using client_id from client certificate

### PR DESCRIPTION
**Pull request by MarcialRosales**
Implements feature request https://github.com/rabbitmq/rabbitmq-server/issues/9323

### CODE REVIEW

File: docs/mqtt.md [hunk 1]
```diff
@@ -594,6 +594,31 @@ to reconnect to switch to a new virtual host.
 is considered more specific than the port-to-vhost mapping with the `mqtt_port_to_vhost_mapping`
 global parameter and so takes precedence over it.
 
+#### Use client_id from client certificate to authenticate 
+
```
Comment (pstack2021) [inline docs/mqtt.md:598]: Please change the title to: Use the client_id from the Client Certificate to Authenticate

I also added capitalization to some of the words in the heading, please check other sub section titles in OSS RabbitMQ topics, if they use capitalisation for 2nd/3rd words in titles then please keep it consistent here, if they don't then you can leave as is, thanks.

File: docs/mqtt.md [hunk 2]
```diff
@@ -594,6 +594,31 @@ to reconnect to switch to a new virtual host.
...
+By the default, the plugin passes the `client_id` it received from the MQTT connection to the configured authentication backends as a credential. However, it is possible to extract the `client_id` from the certificate in a very similar way as we extract the username. 
+
```
Comment (pstack2021) [inline docs/mqtt.md:600]: Perhaps rephrase the 1st sentence to: By the default, the plugin passes the `client_id` as a credential to the configured authentication backends.

Do they need to know this < which it received from the MQTT connection> if not I would leave it out?

Comment (pstack2021) [reply docs/mqtt.md:600]: For the 2nd sentence how about: "However, it is possible to extract the `client_id` from the certificate using the same way the username is extracted. "

Is is the same way or similiar? similar is a bit vague?

Comment (MarcialRosales) [reply docs/mqtt.md:600]: You are right, we may leave that detail out. Thanks!

File: docs/mqtt.md [hunk 3]
```diff
@@ -594,6 +594,31 @@ to reconnect to switch to a new virtual host.
...
+First, you must configure which certificate's attribute has the `client_id` by setting the configuration variable `mqtt.ssl_cert_client_id_from`. The allowed values are:
```
Comment (pstack2021) [inline docs/mqtt.md:601]: What does this mean: "ou must configure which certificate's attribute has the `client_id` " does the user need to search which one is it? 

I would say "The acceptable values are:"

File: docs/mqtt.md [hunk 4]
```diff
@@ -594,6 +594,31 @@ to reconnect to switch to a new virtual host.
...
+- `distinguished_name`, this is the full DN, or distinguished name, of the certificate
```
Comment (pstack2021) [inline docs/mqtt.md:602]: Here I would just say, "this is the full distinguished name, of the certificate." is there a reason for saying "the full DN"?? if not, I would leave it out

File: docs/mqtt.md (hunk 3 above)
Comment (MarcialRosales) [reply docs/mqtt.md:601]: how about: `you must configure where RabbitMQ extract the client_id from the certificate?`

File: docs/mqtt.md [hunk 5]
```diff
@@ -594,6 +594,31 @@ to reconnect to switch to a new virtual host.
...
+- `common_name`, this is just the CN part of the distinguished name
+- `subject_alternative_name` or `subject_alt_name`, here the `client_id` is specified in certificate's extension which allows to specify further alternative names which can be of several types.
```
Comment (pstack2021) [inline docs/mqtt.md:604]: again I would rephrase: ", in these X, the `client_id` is specified in the certificate's extension. You can specify further alternative names which can be several types."

File: docs/mqtt.md (hunk 1 above)
Comment (MarcialRosales) [reply docs/mqtt.md:598]: umm. but the outer title at level 3 does not have that type of capitalization ... `Authentication with TLS/x509 client certificates` ... ??

File: docs/mqtt.md [hunk 6]
```diff
@@ -594,6 +594,31 @@ to reconnect to switch to a new virtual host.
...
+
+If you set `mqtt.ssl_cert_client_id_from` to `subject_alternative_name` or `subject_alt_name`, you should configure `mqtt.ssl_cert_client_id_san_type` which contains the `client_id`, otherwise it defaults to `dns`, its type which can be:
```
Comment (pstack2021) [inline docs/mqtt.md:606]: ?? "you should" or "you must" in the 1st sentence? Do they have to do it, "you should" is too vague, if you are going to use "should", I would additionally explaining why they should, thanks

Comment (pstack2021) [reply docs/mqtt.md:606]: its type which can be:

??? should that be a new sentence written as "It's type can be: "

File: docs/mqtt.md [hunk 7]
```diff
@@ -594,6 +594,31 @@ to reconnect to switch to a new virtual host.
...
+- `dns`
+- `ip` 
+- `email` 
+- `uri`
+- `other_name`
+
+And if there can be more than alternative name of the type specified, you can configure which one to choose via the `mqtt.ssl_cert_client_id_san_index` configuration variable. By default, it always pick the first name, i.e. `mqtt.ssl_cert_client_id_san_index = 0`.
```
Comment (pstack2021) [inline docs/mqtt.md:613]: can't start a sentence with "And" so how about "If there can be additional alternative names of the type that is specified,......

File: docs/mqtt.md [hunk 8]
```diff
@@ -594,6 +594,31 @@ to reconnect to switch to a new virtual host.
...
+
+Here is an example where the username is extracted from the certificate's `common_name` and the `client_id` from the first SAN -subject alternative name- of type uri:
+
```
Comment (pstack2021) [inline docs/mqtt.md:616]: I would say "and the `client_id` is extracted....

File: versioned_docs/version-3.13/mqtt.md [hunk 9]
```diff
@@ -594,6 +594,31 @@ to reconnect to switch to a new virtual host.
 is considered more specific than the port-to-vhost mapping with the `mqtt_port_to_vhost_mapping`
 global parameter and so takes precedence over it.
 
+#### Use client_id from client certificate to authenticate 
+
```
Comment (pstack2021) [inline versioned_docs/version-3.13/mqtt.md:598]: Apply same comments as above here

File: docs/mqtt.md (hunk 6 above)
Comment (MarcialRosales) [reply docs/mqtt.md:606]: I chose `should` because if you do not set it, it is defaulted to `dns`. If that is what the user wants, then the user does not need to configure it.

File: docs/mqtt.md (hunk 7 above)
Comment (MarcialRosales) [reply docs/mqtt.md:613]: I keep forgetting that rule !! :(

File: docs/mqtt.md (hunk 2 above)
Comment (MarcialRosales) [reply docs/mqtt.md:600]: i leave out "the same way " .. to avoid any confusion.

File: docs/mqtt.md (hunk 1 above)
Comment (pstack2021) [reply docs/mqtt.md:598]: ok if the others don't have it, leave as it but include the additional words I suggested, thanks

File: docs/mqtt.md (hunk 6 above)
Comment (pstack2021) [reply docs/mqtt.md]: then I would say "If you want `mqtt.ssl_cert_client_id_from` to  be set to`subject_alternative_name` or `subject_alt_name`, then you must.....

File: docs/mqtt.md (hunk 7 above)
Comment (pstack2021) [reply docs/mqtt.md]: No problem :)

File: docs/mqtt.md (hunk 6 above)
Comment (MarcialRosales) [reply docs/mqtt.md]: At the end, i have gone for "you can configure ".. if the user wants to use `dns` as the SAN type, it is not required to set it `mqtt.ssl_cert_client_id_san_type` to `dns` as that is the default behaviour. Are ok ?
